src/manageiq_client/api.py

```diff
@@ -322,18 +327,25 @@ def options(self, **params):
 
     @property
     def count(self):
-        self.reload_if_needed()
+        self.reload()
```

Comment:
Is this necessary?  This is a property, so every time one accesses `coll.count` it will reload.  Previously it only did it once.

Both ways have benefits in some cases — fresh data vs caching — the questions IMHO are:
1. predictable interface: when one writes/reads code using the client, is it easy to know what operations fetch fresh data and which may be cached?

    The "default" python convention is it's normal for method calls to do heavy operations such as fetching remote data, while attributes/properties are expected to be cheap.  (Let's lump "fetch once then cache" into the cheap category.)

2. performance: is it *possible* to use interface without causing unnecessary queries?  sooner or later someone will care about this, will it require extending the interface?

    Silly example: suppose you're interested in both `coll.count` and `coll.subcount`, now it will necessarily do 2 queries.

Reply:
Makes sense, our usual usecase in tests is like
```python
assert smtg.count == 0
smtg.action.create(...)
smtg.reload()
assert smtg.count == 1
```
so my idea was to simplify it, but your explanation with properties vs methods makes sense.

I'm still inclined to do full ``reload()`` in case of ``.all``, even though it's property. It proved very confusing that e.g.
```python
for resource in collection.all
```
gives different results than
```python
for resource in collection
```
What do you think?

Reply:
I'd like to map out the big picture before having an opinion...

Indeed, confusing, but should `__iter__` do reload?
Hmm, a Collection (and other classes too esp. Entity) serves double duty: 
1. a "proxy" denoting the actual collection in ManageIQ.
2. a "result" holding specific data that has been retrieved

Yeah, I agree that `for x in client.collections.foos:` is high-level code that uses the first "proxy" sense, it better fetch current data.  (And therefore so should `for x in client.collections.foos.all:`.)

I wonder if for the "result" sense, instead of occasionally mutating the collection/entity, we could have separate classes.  You could then retain the result, examine it as much as wanted, and when desired obtain a new result...
`SearchResult` already is that, for one use case.  And the return result of `Collection.all` — simply a list of `Entity`s — is a representation of current collection result.
Not sure this leads anywhere good, just a thought I had...